### PR TITLE
Enable new PyTorch autocast

### PR DESCRIPTION
**Pull request by emcastillo**

### CODE REVIEW

File: pytorch_pfn_extras/utils/__init__.py [hunk 1]
```diff
@@ -1,2 +1 @@
 from pytorch_pfn_extras.utils import checkpoint  # NOQA
-from pytorch_pfn_extras.utils import comparer  # NOQA
```
Comment (emcastillo) [inline pytorch_pfn_extras/utils/__init__.py:2]: avoid circular deps when importing utils module from logic or runtime

File: pytorch_pfn_extras/handler/__init__.py [hunk 2]
```diff
@@ -1,6 +1,3 @@
 from pytorch_pfn_extras.handler._code_block import CodeBlock, update_parameters, forward  # NOQA
 from pytorch_pfn_extras.handler._handler import BaseHandler, Handler  # NOQA
 from pytorch_pfn_extras.handler._logic import BaseLogic, Logic, CodeBlockLogic  # NOQA
-
-# Deprecated, only imported for backward compatibility
-from pytorch_pfn_extras.handler._logic import torch_autocast  # NOQA
```
Comment (kmaehashi) [inline pytorch_pfn_extras/handler/__init__.py:6]: Maybe keep it as is?

File: pytorch_pfn_extras/handler/_logic.py [hunk 3]
```diff
@@ -179,7 +162,14 @@ def __init__(
                     the gradient.
                 * ``'autocast'`` (bool):
                     If ``True``, ``torch.cuda.amp.autocast`` is enabled.
-                    Default is ``False``.
+                    Default is ``False``. This is deprecated in favor of
+                    ``autocast_options``.
+                * ``'autocast_options'`` (bool):
```
Comment (kmaehashi) [inline pytorch_pfn_extras/handler/_logic.py:167]: How about accepting `autocast={'dtype': torch.bfloat16}` instead of adding a new option?

**Review by kmaehashi — COMMENTED** [review]
Thanks! Reviewed the API surface.

File: pytorch_pfn_extras/handler/__init__.py (hunk 2 above)
Comment (emcastillo) [reply pytorch_pfn_extras/handler/__init__.py:6]: I think it may be better to cut this at some point if possible :)

Comment (emcastillo) [reply pytorch_pfn_extras/handler/__init__.py]: Reverted!

File: pytorch_pfn_extras/handler/_logic.py (hunk 3 above)
Comment (emcastillo) [reply pytorch_pfn_extras/handler/_logic.py:179]: Done!

**Comment by emcastillo** [comment]
PTAL!

File: pytorch_pfn_extras/runtime/_autocast.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                or self._options["device_type"] == "cuda"
+            ):
+                raise RuntimeError('Requested AMP features but torch.cuda.amp'
+                                   ' is not enabled')
+
+    @contextlib.contextmanager
+    def autocast(self, enabled: bool = True) -> Generator[None, None, None]:
+        # CUDA Availability was checked in Runtime Constructor
+        if self._use_old_ac:
+            with torch.cuda.amp.autocast(self._options["enabled"]):  # type: ignore[no-untyped-call]
```
Comment (kmaehashi) [inline pytorch_pfn_extras/runtime/_autocast.py:52]: How about passing `**self._options`? The arguments are compatible:
https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.autocast

**Comment by emcastillo** [comment]
PTAL!!

File: pytorch_pfn_extras/runtime/_autocast.py (hunk 4 above)
Comment (emcastillo) [reply pytorch_pfn_extras/runtime/_autocast.py]: torch.cuda.amp.autocast

`device_type` is not compatible :(
https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.autocast

File: pytorch_pfn_extras/handler/_logic.py [hunk 5]
```diff
@@ -177,9 +171,16 @@ def __init__(
                 * ``'backward_outputs'`` (list of str):
                     A list of names of outputs that require compution of
                     the gradient.
-                * ``'autocast'`` (bool):
+                * ``'autocast'`` (bool or dict):
                     If ``True``, ``torch.cuda.amp.autocast`` is enabled.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/handler/_logic.py:175]:
```suggestion
                    If ``True``, ``torch.autocast`` (or ``torch.cuda.amp.autocast`` for PyTorch 1.9 or earlier) is enabled,
```

File: pytorch_pfn_extras/handler/_logic.py [hunk 6]
```diff
@@ -177,9 +171,16 @@ def __init__(
...
-                    Default is ``False``.
+                    using ``{"enabled": True, "device_type": "cuda"}``
+                    as autocast options.
+                    Default is ``False`` which corresponds to the following options
+                    ``{"enabled": False, "device_type": "cuda"}``
+                    dict type. If dict, Options to pass to ``torch.autocast``.
+                    Includes ``device_type``, ``dtype`` among others.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/handler/_logic.py:181]:
```suggestion
                    The default is ``False`` which corresponds to the following options
                    ``{"enabled": False, "device_type": "cuda"}``.
                    If dict, options are passed to ``torch.autocast``.
```

File: pytorch_pfn_extras/runtime/_autocast.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    pass
+
+
+class _AutocastManager:
+    def __init__(
+        self,
+        autocast_options: Dict[str, Any],
+        has_grad_scaler: bool,
+    ) -> None:
+        self._options = autocast_options
```
Comment (kmaehashi) [inline pytorch_pfn_extras/runtime/_autocast.py:21]:
```suggestion
        autocast_options = autocast_options.copy()
        self._enabled = autocast_options.pop('enabled', True)
        self._device_type = autocast_options.pop('device_type', 'cuda')
        self._options = autocast_options
```

File: pytorch_pfn_extras/runtime/_autocast.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        self._use_old_ac = not requires("1.10.0")
+        if (
+            self._use_old_ac
+            and self._options.get("enabled", False)
+            and self._options.get("device_type", "cuda") != "cuda"
```
Comment (kmaehashi) [inline pytorch_pfn_extras/runtime/_autocast.py:26]:
```suggestion
            self._enabled and self._use_old_ac and self._device_type == 'cuda'
```

File: pytorch_pfn_extras/handler/_logic.py [hunk 9]
```diff
@@ -177,9 +171,16 @@ def __init__(
...
+            * ``'grad_scaler'`` (torch.cuda.amp.GradScaler):
+                A gradient scaler that outputs are applied to.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/handler/_logic.py:183]:
```suggestion
```

File: pytorch_pfn_extras/runtime/_autocast.py (hunk 4 above)
Comment (kmaehashi) [reply pytorch_pfn_extras/runtime/_autocast.py:42]:
```suggestion
            with torch.cuda.amp.autocast(enabled=self._enabled, **self._options):  # type: ignore[no-untyped-call]
```

File: pytorch_pfn_extras/runtime/_autocast.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+                                   ' is not enabled')
+
+    @contextlib.contextmanager
+    def autocast(self, enabled: bool = True) -> Generator[None, None, None]:
+        # CUDA Availability was checked in Runtime Constructor
+        if self._use_old_ac:
+            with torch.cuda.amp.autocast(self._options["enabled"]):  # type: ignore[no-untyped-call]
+                yield
+        else:
+            with torch.autocast(**self._options):  # type: ignore[no-untyped-call,attr-defined]
```
Comment (kmaehashi) [inline pytorch_pfn_extras/runtime/_autocast.py:45]:
```suggestion
            with torch.autocast(self._device_type, enabled=self._enabled, **self._options):  # type: ignore[no-untyped-call,attr-defined]
```

**Review by kmaehashi — COMMENTED** [review]
Maybe like this?

**Review by kmaehashi — APPROVED** [review]
Thanks, LGTM!

**Pull request merged**